lib/puppet/util/autoload.rb

```diff
@@ -67,6 +69,12 @@ def load_file(name, env)
         message = "Could not autoload #{name}: #{detail}"
         Puppet.log_exception(detail, message)
         raise Puppet::Error, message, detail.backtrace
+      ensure
+        # Restore original LOAD_PATH
+        $LOAD_PATH.each do |p|
+          $LOAD_PATH.delete(p) unless load_path.include? p
+        end
+        $LOAD_PATH.uniq!
```

Comment:
I agree this isn't an ideal way to do it.  I'm not entirely sure what's
already on LOAD_PATH (apart from itself) - certainly
module_directories(env) needs adding.

In terms of the restore, my preference would have been to set $LOAD_PATH
directly, but it's not a normal array (you can't directly set it).

Though it's not ideal, it doesn't seem to be causing any performance issues
on my system.  The first instance of a match short-circuits the rest of the
array.  It's the exceptional case where something doesn't exist that will
take the worst hit.

What might be a more suitable solution would be to prepare LOAD_PATH before
compilation (but only when the real client environment is known), and then
drop it back to its previous state after compilation is finished.

I'd gladly accept suggestions of how to achieve this (i.e. where this could
reasonably be hooked in) and update my PR accordingly.

On 13 November 2015 at 13:47, Henrik Lindberg <notifications@github.com>
wrote:

> In lib/puppet/util/autoload.rb
> <https://github.com/puppetlabs/puppet/pull/4445#discussion_r44783036>:
>
> > @@ -67,6 +69,12 @@ def load_file(name, env)
> >          message = "Could not autoload #{name}: #{detail}"
> >          Puppet.log_exception(detail, message)
> >          raise Puppet::Error, message, detail.backtrace
> > +      ensure
> > +        # Restore original LOAD_PATH
> > +        $LOAD_PATH.each do |p|
> > +          $LOAD_PATH.delete(p) unless load_path.include? p
> > +        end
> > +        $LOAD_PATH.uniq!
>
> The above restore will have performance impact. It should be possible to
> do this in a more light way fashion. The above has two linear array
> searches for each entry on the path + the extra uniq set operation at the
> end.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppet/pull/4445/files#r44783036>.
>



-- 


Nick Moriarty
Linux Systems Administrator and Developer

Information Services
University of York
Heslington
York
YO10 5DD
+44 (0)1904 32 3484



e-mail disclaimer: http://www.york.ac.uk/docs/disclaimer/email.htm